Improve common time and date crates support

Anfid opened the pull request:
Fixes: #741

## Goals

Goal of this MR is to improve the support of common time crates.

## Changes

Breaking changes were introduced and a major version bump would be required.

* Reworked `scylla::frame::value` date and time types:
  * Align structs with internal CQL representation;
  * Rename to add `Cql`- prefix to explicitly show direct relation to CQL types;
* Moved `chrono` dependency of `scylla-cql` behind `scylla-cql/chrono` feature;
* Added optional `time` dependency with full serialization support;
* Implemented traits to convert between chrono/time types and CQL types;
* Removed implementation of `FromCqlValue<CqlValue>` for `chrono::Duration` as it may be misleading. `CqlTime` and `CqlTimestamp` structs are no longer a `chrono::Duration` wrappers;
* `CqlValue::as_date` function signature changed to return `time::Date` instead of `chrono::NaiveDate` for better naming consistency;
* `CqlValue::as_cql_time` and `CqlValue::as_cql_timestamp` functions now ensure that row is of Time or Timestamp type correspondingly. Upstream behavior is to parse all bignum types as time or timestamp;

## Points for discussion

1) If we can guarantee that CQL time type never exceeds 23:59:59.999,999,999 extra guarantees can be given about `CqlTime` struct and implementations of `TryFrom<CqlTime>` for `chrono::NaiveTime` and `time::Time` can be replaced with `From<CqlTime>`. However, this would require restricting access to internal value of `CqlTime` (like implementing constructor and getter). I wasn't sure if we can give this guarantee, so it's not present in current implementation;
2) Support for `std::time::SystemTime` was intentionally not implemented as it represents local time and not UTC time. This may and probably will lead to incorrect values in the database tables, as CQL timestamp is UTC;
3) `scylla::history` module requires `chrono` dependency for `scylla` crate. I wasn't sure about how to best handle it nor its intended use case, so I didn't touch it. Potentially it could be nice to change `history::TimePoint` type based on what features were used.

## Other concerns

I wasn't able to run `test/dockerized/run.sh` script without changes. On my machine no containers were found with `grep cluster_scylla` command on line 5. It worked after replacing it with `grep cluster-scylla` locally. I would need more info on behavior on machines other than mine to push this fix.

This MR will produce a lot of conflicts with #665, I'll tag here it so it won't go unnoticed.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [x] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

## Pre-merge checklist

- [x] Set up CI for feature flags combinations.

I would appreciate some help with setting up CI for feature flag combinations as added functionality is NOT tested with github CI.

wprzytula commented on scylla/src/transport/cql_types_test.rs:
```diff
@@ -295,108 +293,189 @@ async fn test_naive_date() {
 #[tokio::test]
-async fn test_time() {
+async fn test_cql_time() {
     // Time is an i64 - nanoseconds since midnight
     // in range 0..=86399999999999
```
`Time` -> `CqlTime`

wprzytula commented on scylla/src/transport/cql_types_test.rs:
```diff
@@ -406,19 +485,19 @@ async fn test_time() {
         // Insert time as a bound Time value and verify that it matches
```
`Time` -> `CqlTime`

wprzytula commented on scylla/src/transport/cql_types_test.rs:
```diff
@@ -510,19 +737,19 @@ async fn test_timestamp() {
         // Insert timestamp as a bound Timestamp value and verify that it matches
```
`Timestamp` -> `CqlTimestamp`

wprzytula reviewed (changes requested):
Great job! This is a very valuable PR.
Only small nits here, mainly typos and old naming leftovers. Wonderful to have the time dependencies separated with a feature, and with proper test coverage.
Will you be able to address the comments?

Thanks a lot for contributing.

wprzytula commented:
> I would appreciate some help with setting up CI for feature flag combinations as added functionality is NOT tested with github CI.

How can we help you? Would you like us to provide a code snippet to put in CI config files that sets up tests for feature flag combinations?

Anfid commented:
> Would you like us to provide a code snippet to put in CI config files that sets up tests for feature flag combinations?

Yes, that would help a lot. I believe you also have the permission to push commits directly to this MR, so just pushing an extra commit is also an option, if that's easier for you.

Alternatively, you could just say which feature combinations you want tested, but it may take me extra time to implement it in GitHub CI, as I'm not super familiar with it. My main concern was that testing feature flag combinations extensively may get quite costly, as each flag combination would require a recompilation and might mess up with cache. I wasn't able to decide on where would be a good balance between test exhaustiveness and CI execution time for this project, so I just left a comment asking for some help with it instead.

wprzytula commented:
> Yes, that would help a lot. I believe you also have the permission to push commits directly to this MR, so just pushing an extra commit is also an option, if that's easier for you.

Let's do so. Once you make the corrections I requested, I will add a commit with CI-related changes.

Anfid commented on scylla-cql/src/frame/value.rs:
```diff
@@ -41,20 +42,206 @@ pub enum MaybeUnset<V: Value> {
 
 /// Wrapper used to differentiate between Time and Timestamp as sending values
 /// Nanoseconds since midnight
 #[derive(Clone, Copy, PartialEq, Eq, Debug)]
```
Since original docstrings don't make a lot of sense with this implementation, I decided to rewrite the entire docstring, rather than just update the type name

Anfid commented on scylla-cql/src/frame/response/cql_to_rust.rs:
```diff
@@ -162,7 +186,27 @@ impl FromCqlVal<CqlValue> for crate::frame::value::Time {
 
-impl FromCqlVal<CqlValue> for crate::frame::value::Timestamp {
+#[cfg(feature = "chrono")]
+impl FromCqlVal<CqlValue> for chrono::NaiveTime {
```
Not really, I was just a bit hesitant to introduce another import hidden behind a feature flag. Fixed by importing `chrono::NaiveTime` for consistency

Anfid commented on scylla-cql/src/frame/response/cql_to_rust.rs:
```diff
@@ -171,13 +215,23 @@ impl FromCqlVal<CqlValue> for crate::frame::value::Timestamp {
+        CqlTimestamp(timestamp)
+            .try_into()
+            .map_err(|_| FromCqlValError::BadVal)
     }
```
Certainly, not sure why I was hesitant to type check them too in the original commit

Anfid commented:
Finally got some time to work on corrections, everything should be resolved now. After resolving conflicts with `main`, some unexpected issues came up in `scylla/src/utils/pretty.rs`. Math in formatting has to be reworked, so I'll work on that shortly.

Meanwhile, could you please also check if issue with `test/dockerized/run.sh` script I described is something local to my machine (macOS) or is it an issue that should be fixed for everyone? I can also push an update for this script, since I have already updated it locally to run the tests.

Anfid commented:
Conflicts with the base branch were resolved, please also see additional changes to `scylla/src/utils/pretty.rs`. I decided to keep using chrono for rendering time types.

It should also be possible to remove `scylla`'s dependency on `chrono`, but I believe it to be out of scope of this PR, because it would require a lot of other changes unrelated to time crates support.

piodul reviewed (changes requested):
Sorry for the long delay, I'll take over the review now and will give it more concentrated attention.

Thanks, I like the direction of the PR very much. I like the idea of `CqlTime`/`CqlTimestamp`/`CqlDate` which are independent from external crates. I appreciate comprehensive tests, implementations of conversion traits and the updated documentation (folks usually forget about the last one).

It's a lot of changes introduced at once, so I hope I didn't miss anything, but the implementation looks OK. I left a small number of asks, the biggest ones being:

1. Use the new `CqlXYZ` types in the `CqlValue`. This will require mechanical changes in multiple places, but I believe this should simplify some of the code so that manually wrapping/unwrapping from/to these types is not needed.
2. Make the new `CqlXYZ` types more hermetic and provide a constructor + getter for them. This would be nice to have but I think can be done in a followup.


> ## Points for discussion
> 1. If we can guarantee that CQL time type never exceeds 23:59:59.999,999,999 extra guarantees can be given about `CqlTime` struct and implementations of `TryFrom<CqlTime>` for `chrono::NaiveTime` and `time::Time` can be replaced with `From<CqlTime>`. However, this would require restricting access to internal value of `CqlTime` (like implementing constructor and getter). I wasn't sure if we can give this guarantee, so it's not present in current implementation;

I like the idea to restrict the access to the internal field of `CqlTime`/`CqlTimestamp`/`CqlDate`. I'm not sure what are your concerns about not being able to provide the guarantees, but to me it looks like we just need to do as you wrote: make the inner field private, introduce a constructor and a getter. Of course, the rest of the module that the types are defined in will have access to the internal field but you can reduce the risk of manipulating on the type directly by moving it to a dedicated module and then re-exporting it back in `value.rs`.

That said, I don't consider it a blocker for merging this PR, we can leave it for a followup if you want.

> 2. Support for `std::time::SystemTime` was intentionally not implemented as it represents local time and not UTC time. This may and probably will lead to incorrect values in the database tables, as CQL timestamp is UTC;

Sure, makes perfect sense.

> 3. `scylla::history` module requires `chrono` dependency for `scylla` crate. I wasn't sure about how to best handle it nor its intended use case, so I didn't touch it. Potentially it could be nice to change `history::TimePoint` type based on what features were used.

@cvybhu do you remember `scylla::history` uses `chrono`? AFAIR the history module was supposed to allow to track what the driver was doing during the query; I can see some time points being created from `SystemTime::now()` - would it be possible to switch to using `SystemTime` for that? Additionally, perhaps it makes sense to use `Instant`s instead of `SystemTime` as the latter is not monotonic (https://doc.rust-lang.org/std/time/struct.SystemTime.html).

----

As for the CI checks, as the `time`/`chrono` features seem pretty orthogonal, I think it's sufficient to test whether the crate compiles with and without all serialization-related features, and only run tests with all serialization features enabled.

I pushed some commits to your branch that modify the existing Github workflows. The CI will most likely fail because of the new version of clippy that started to complain about some existing code (please rebase), but it should be a good starting point. They should work after rebase, but I can't tell for sure because I didn't test it - I'll correct them if there are any issues after rebase.

Anfid commented on scylla/src/utils/pretty.rs:
```diff
@@ -261,15 +260,12 @@ mod tests {
+        let t = chrono::NaiveTime::from_hms_nano_opt(6, 5, 4, 123)
+            .unwrap()
+            .signed_duration_since(chrono::NaiveTime::MIN);
+        let t = t.num_nanoseconds().unwrap();
```
I'd love to, but this would require enabling `chrono` feature for all tests, which is probably not the desired behavior

piodul replied:
I see. In that case it's fine.

Anfid commented on scylla-cql/src/frame/response/cql_to_rust.rs:
```diff
@@ -144,16 +144,40 @@ impl<const N: usize> FromCqlVal<CqlValue> for [u8; N] {
+            _ => Err(FromCqlValError::BadCqlType),
+        }
+    }
+}
```
I wanted to avoid the `Option` indirection in the implementation, but this shouldn't matter that much. Updated the code according to the suggestion

Anfid commented:
Sorry for the huge delay on my side too :)

I've rebased the changes. There were a couple of conflicts I had to resolve, you might want to check those parts of code. In particular, there were conflicts with `.github` workflows, I've updated your implementation to run against `--all-targets`, as it is currently in upstream.

There was an issue with 2 of my unit tests where I used `#[cfg(features = "chrono")]` instead of `#[cfg(feature = "chrono")]`. Luckily updated Rust compiler was able to catch it and I fixed the tests accordingly.

All CQL time types in the CqlValue type were changed to wrapper types, as you've suggested. I've left their internal fields public for now, I'll try to update it in the follow-up PR.

> I like the idea to restrict the access to the internal field of CqlTime/CqlTimestamp/CqlDate. I'm not sure what are your concerns about not being able to provide the guarantees

My concerns are the leap second representation within the internal database CQL type. I'm not sure if it's supported, so I'd prefer to see any explicit documentation on that before giving any guarantees to the users of the library. Regarding restricting access to the internal fields of wrapper types. I really like that idea, but I'd prefer to do it in a separate follow-up PR to avoid further delays.

piodul commented:
> All CQL time types in the CqlValue type were changed to wrapper types, as you've suggested. I've left their internal fields public for now, I'll try to update it in the follow-up PR.

:+1:

> > I like the idea to restrict the access to the internal field of CqlTime/CqlTimestamp/CqlDate. I'm not sure what are your concerns about not being able to provide the guarantees
> 
> My concerns are the leap second representation within the internal database CQL type. I'm not sure if it's supported, so I'd prefer to see any explicit documentation on that before giving any guarantees to the users of the library. Regarding restricting access to the internal fields of wrapper types. I really like that idea, but I'd prefer to do it in a separate follow-up PR to avoid further delays.

I see. This is from the [protocol spec](https://github.com/scylladb/scylla-tools-java/blob/86a200e324fac3b4a47b7954e30de6e05a4e7ce0/doc/native_protocol_v4.spec#L941C1-L944):

```
6.16 time

  An 8 byte two's complement long representing nanoseconds since midnight.
  Valid values are in the range 0 to 86399999999999
```

It looks like leap seconds are not supported, so - provided that the internal field of CqlTime is made private - maybe it would make sense to change the conversion from `CqlTime` to `chrono::NaiveTime` to be infallible, but not the other way around. It's fine to move this to the followup which makes the internal fields private.

I'll re-review the PR now.

piodul commented:
> It looks like leap seconds are not supported, so - provided that the internal field of CqlTime is made private - maybe it would make sense to change the conversion from `CqlTime` to `chrono::NaiveTime` to be infallible, but not the other way around. It's fine to move this to the followup which makes the internal fields private.

Sigh... Specification is one thing, but implementation is another. It looks like Scylla happily accepts an out-of-range `time` value.

Although we already have an 'if' on a deserialization path which returns an error when out-of-range values are returned from the database, I think it should be removed to take account for incorrectly inserted values into the database. Likewise, I take back what I said about the conversion from CqlTime to NaiveTime - it should be fallible.

piodul reviewed (commented):
Looks good! The only thing that needs fixing is the lockfile in the MSRV check. As you modified the Cargo.toml files, you need to regenerate the Cargo.toml.msrv.

Our current MSRV is 1.66. Last time the lockfile was regenerated, all the newest versions of the dependencies worked with it except for the `toml_datetime` indirect dependency, but `time` may pull in some other problematic dependencies.

AFAIK the process is:

- Generate the Cargo.lock locally
- Pin versions for deps that won't compile in 1.66 to the last version which works, for `toml_update` it would be `cargo update -p toml_datetime --precise 0.6.3`
- Rename to `Cargo.lock.msrv`

@Lorak-mmk please correct me if something is wrong in the above procedure. By the way - we should either document it or make it more convenient to generate `Cargo.toml.msrv` (a script maybe?).

Anfid commented:
Done. It seems that a lot of dependencies can be updated without breaking MSRV, but I decided to avoid being intrusive with changes unrelated to the purpose of this PR and pinned all versions to whatever was in the previous lockfile

Anfid commented:
Oh, my bad, I'll fix it shortly

piodul commented:
(Forgot to squash before merging previously, I dropped the merge commit and pushed a squashed version)
